src/riscv/context.rs

```diff
@@ -170,13 +232,92 @@ impl TaskContext {
             unsafe { crate::asm::write_user_page_table(next_ctx.satp) };
             crate::asm::flush_tlb(None); // currently flush the entire TLB
         }
-        unsafe {
-            // TODO: switch FP states
-            context_switch(self, next_ctx)
+        #[cfg(feature = "fp-simd")]
+        {
+            use riscv::register::sstatus;
+            use riscv::register::sstatus::FS;
+            // get the real FP state of the current task
+            let current_fs = sstatus::read().fs();
+            // save the current task's FP state
+            if current_fs == FS::Dirty {
+                // we need to save the current task's FP state
+                unsafe {
+                    save_fp_registers(&mut self.fp_status.fp);
+                }
+                // after saving, we set the FP state to clean
+                self.fp_status.fs = FS::Clean;
+
+                // Synchronize TrapFrame's FS state
+                #[cfg(feature = "uspace")]
+                if let Some(trapframe) = unsafe { self.get_trapframe_mut() } {
+                    // Clear FS bits and set to Clean
+                    trapframe.sstatus &= !(0x6000); // Clear FS bits (bit 13-14)
+                    trapframe.sstatus |= (FS::Clean as usize) << 13; // Set FS=Clean
+                }
+            }
+            // restore the next task's FP state
+            match next_ctx.fp_status.fs {
+                FS::Clean => unsafe {
+                    // the next task's FP state is clean, we should restore it
+                    restore_fp_registers(&next_ctx.fp_status.fp);
+                    // after restoring, we set the FP state
+                    sstatus::set_fs(FS::Clean);
```

Comment:
Read and write `sstatus` via `TrapFrame`, do not access hardware register directly.

Please refer to Linux's implementation.